Update to revision af28b719e6 of Core [ECR-4046]

vitvakatu opened the pull request:
## Overview

Some kind of merge conflict occured, a lot of commits are actually from a different PR, but overall diff is correct.

---
See: https://jira.bf.local/browse/ECR-4046

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [ ] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [ ] Public API has Javadoc
- [ ] Method preconditions are checked and documented in the Javadoc of the method
- [ ] Changelog is updated if needed (in case of notable or breaking changes)
- [ ] The continuous integration build passes

vitvakatu commented on exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceRuntimeAdapterTest.java:
```diff
@@ -123,7 +126,7 @@ void addService() throws CloseFailuresException {
     ServiceInstanceSpec expected = ServiceInstanceSpec.newInstance(serviceName, serviceId,
-        ServiceArtifactId.newJavaId(javaArtifactName));
+        ServiceArtifactId.newJavaId(artifactId));
     verify(serviceRuntime).startAddingService(fork, expected, configuration);
```
I can't understand why this test isn't working.
```
[ERROR] Failures: 
[ERROR]   ServiceRuntimeAdapterTest.addService:130 
Argument(s) are different! Wanted:
serviceRuntime.startAddingService(
    com.exonum.binding.core.storage.database.Fork@76c12d09,
    ServiceInstanceSpec{name=s1, id=1, artifactId=1:com.acme:foo:1.2.3},
    [(byte) 0x01, (byte) 0x02]
);
-> at com.exonum.binding.core.runtime.ServiceRuntime.startAddingService(ServiceRuntime.java:189)
Actual invocations have different arguments:
serviceRuntime.startAddingService(
    com.exonum.binding.core.storage.database.Fork@76c12d09,
    ServiceInstanceSpec{name=s1, id=1, artifactId=1:com.acme:foo:1.2.3},
    [(byte) 0x01, (byte) 0x02]
);
-> at com.exonum.binding.core.runtime.ServiceRuntimeAdapter.startAddingService(ServiceRuntimeAdapter.java:131)
```

dmitry-timofeev replied:
Their _string representation_ is identical; not the objects (artifactId most certainly — I suspect one has name 'com.acme' + version 'foo:1.2.3', the other — normal).

dmitry-timofeev commented on exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceArtifactId.java:
```diff
@@ -25,29 +25,34 @@
- * deployed and the service artifact name. The name of Java artifacts usually contains the three
- * coordinates identifying any Java artifact: groupId, artifactId and version.
+ * deployed, the service artifact name and the artifact version. The name of Java
+ * artifacts usually consists of groupId and artifactId.
```
artifacts usually consists of the two Maven coordinates: groupId and artifactId.

dmitry-timofeev commented on exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceArtifactId.java:
```diff
@@ -25,29 +25,34 @@
 
   /**
-   * Returns the full artifact name of this service (e.g., "com.acme:land-registry:1.2.0").
+   * Returns the full artifact name of this service (e.g., "com.acme:land-registry").
```
You can't expect the rest of the code to work by using ':' as delimiter for groupId and artifactId.

dmitry-timofeev replied:
'full' no longer makes sense.

dmitry-timofeev commented on exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceArtifactId.java:
```diff
@@ -25,29 +25,34 @@
+  /**
+   * Parses a service id in format "runtimeId:serviceName:version" as {@link #toString()} produces.
    *
    * @param serviceArtifactId a string in format "runtimeId:serviceName". Whitespace
```
Docs.

dmitry-timofeev commented on exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceArtifactId.java:
```diff
@@ -58,16 +63,20 @@ public static ServiceArtifactId parseFrom(String serviceArtifactId) {
    */
-  public static ServiceArtifactId newJavaId(String name) {
-    return valueOf(JAVA.getId(), name);
+  public static ServiceArtifactId newJavaId(String artifactName) {
```
It makes no sense: it accepts the _name_, which is `'groupId<delimiter>artifactId'`
  — where from do you pull the version? There must not be 'artifactName' that is not the same as ArtifactId#getName.

Second, the delimiter.

dmitry-timofeev commented on exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceArtifactId.java:
```diff
@@ -76,16 +85,18 @@ public static ServiceArtifactId newJavaId(String name) {
    * @param runtimeId the runtime id in which the service shall be deployed
    * @param name the name of the service; must not be blank
```
docs for version

dmitry-timofeev commented on exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceArtifactId.java:
```diff
@@ -76,16 +85,18 @@ public static ServiceArtifactId newJavaId(String name) {
    */
-  public static ServiceArtifactId valueOf(int runtimeId, String name) {
+  public static ServiceArtifactId valueOf(int runtimeId, String name, String version) {
+    // TODO: lol, error message have to be super useful
```
? blank != empty

dmitry-timofeev commented on exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceArtifactIdTest.java:
```diff
@@ -26,9 +26,9 @@
+      "0:land-registry:1.1",
       "1:com.acme:foo-service:1.1.1-beta1",
-      "100500:::",
+      "100500:not-blank:version",
```
?

dmitry-timofeev commented on exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceArtifactIdTest.java:
```diff
@@ -26,9 +26,9 @@
-      "100500:::",
+      "100500:not-blank:version",
   })
   void parseFromRoundtrip(String serviceId) {
```
Must better be the other way around:
`ServiceArtifactId -- toString --> String -- parseFrom --> ServiceArtifactId* -> assertEquals`

dmitry-timofeev commented on exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceArtifactIdTest.java:
```diff
@@ -41,25 +41,30 @@ void parseFromRoundtrip(String serviceId) {
   void newJavaId() {
-    String name = "test-service";
-    ServiceArtifactId id = ServiceArtifactId.newJavaId(name);
+    String serviceName = "test-service";
```
That is not a usual Java service name — it must include groupId, separated with some other delimiter.

dmitry-timofeev commented on exonum-java-binding/core/rust/src/utils/jni_cache.rs:
```diff
@@ -216,9 +216,9 @@ pub mod runtime_adapter {
     }
 
     /// Returns cached `JMethodID` for `ServiceRuntimeAdapter.commitService()`.
```
docs

dmitry-timofeev commented on exonum-java-binding/core/rust/src/testkit/mod.rs:
```diff
@@ -95,16 +96,14 @@ pub extern "system" fn Java_com_exonum_binding_testkit_TestKit_nativeCreateTestK
             }
 
-            if let Some(service) =
+            if let Some(params) =
```
params does not work

dmitry-timofeev commented on exonum-java-binding/core/rust/src/testkit/mod.rs:
```diff
@@ -95,16 +96,14 @@ pub extern "system" fn Java_com_exonum_binding_testkit_TestKit_nativeCreateTestK
                     .with_artifact(artifact_id)
-                    .with_instance(instance);
+                    .with_instance(&params)
+                    .with_rust_service(params.factory);
```
It is very weird: Why doesn't testkit accept service factory in `with_artifact`?
Why on top of artifact and instance definition a factory is added with a _separate_
and _deceivingly named_ method (it says 'service' but accepts a factory of services — why isn't it passed with artifact)?

dmitry-timofeev commented on exonum-java-binding/core/rust/src/testkit/mod.rs:
```diff
@@ -257,7 +256,7 @@ fn time_service_instance_from_java(
     executor: Executor,
     time_service_spec: JObject,
-) -> JniResult<Option<InstanceCollection>> {
+) -> JniResult<Option<TimeServiceInstanceParams>> {
```
Documentation above has not been updated.

dmitry-timofeev commented on exonum-java-binding/core/rust/src/testkit/mod.rs:
```diff
@@ -288,3 +291,21 @@ fn get_field_as_string(env: &JNIEnv, obj: JObject, field_name: &str) -> JniResul
+    pub service_name: String,
+}
+
+impl<'a> Into<InstanceInitParams> for &'a TimeServiceInstanceParams {
```
Is it a natural use of `Into` — a reference is converted, not a value, and factory
is not actually used in that conversion (just as an artifact_id carrier)?

dmitry-timofeev commented on exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntime.java:
```diff
@@ -211,7 +213,8 @@ public void startAddingService(Fork fork, ServiceInstanceSpec instanceSpec,
    *     is not deployed
    */
-  public void commitService(ServiceInstanceSpec instanceSpec) {
+  public void updateInstanceState(ServiceInstanceSpec instanceSpec,
```
docs

dmitry-timofeev commented on exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/ServiceRuntimeAdapter.java:
```diff
@@ -138,19 +139,21 @@ void startAddingService(long forkHandle, byte[] instanceSpec, byte[] configurati
+   * @see ServiceRuntime#updateInstanceState(ServiceInstanceSpec, InstanceState.Status)
    */
-  void commitService(byte[] instanceSpec) {
+  void updateServiceState(byte[] instanceSpec, int numericInstanceStatus) {
```
docs for numericInstanceStatus

vitvakatu commented:
Reverted update to latest Core revision.
The list of current changes:
- `startAddingService` -> `initiateAddingService`
- `commitService` -> `updateInstanceStatus`. Added new parameter `instanceStatus`. Documentation altered slightly and requires more work.
- Actual protobuf sources for the revision

dmitry-timofeev commented on exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/ServiceRuntimeIntegrationTest.java:
```diff
@@ -257,7 +257,7 @@ void commitService() {
 
     // Create the service from the artifact
-    serviceRuntime.commitService(instanceSpec);
+    serviceRuntime.updateInstanceStatus(instanceSpec, null);
```
What is expected to be in place for `null`? An enum value? If that's the case, I think it would be more appropriate than `null`, even if it is unused.

dmitry-timofeev commented:
The error is in Coveralls. It also shows that the OrderedMapOuterClass is in incorrect package — core needs to be updated.
